Introduce the native_sim board

aescolar opened the pull request:
This PR corresponds to:
https://github.com/zephyrproject-rtos/zephyr/issues/58305

In this PR the following is done:
* Introduce the native simulator (runner) in tree(^1, see #59497)
* Introduce a new board (native_sim & native_sim_64), which use this runner.
* Introduce 2 new kconfig options: NATIVE_BUILD and NATIVE_LIBRARY apart from NATIVE_APPLICATION (application is kept for those that build monolitically, i.e. native_posix & nrf52_bsim), NATIVE_BUILD is common to both NATIVE_LIBRARY and APPLICATION
* For libC , and the POSIX API shim, minor changes to their kconfig and their tests' filters to allow both in this new target.
* Add a few overlays for tests and update tests filters

What comes *after* this PR:
* More tests' yaml filter updates
* Refactoring of more drivers for the POSIX arch in tops and bottoms, so they can be used in native_sim with the embedded libCs.
  A demo of a very simple driver refactor can be found [here](https://github.com/aescolar/zephyr/commit/476f96f954354f27f3b7c7d996d23fae6f425fff), and a more elaborate one in [here](https://github.com/aescolar/zephyr/commit/76b18af90f1206de9c9dcb5cd3701efec64f7cdf)
  Note in the documentation ( boards/posix/native_sim/doc/index.rst )  you can find a list of which drivers are already supported with all C libraries, and which only with the host libC.

----

#### Note (^1)

The native simulator is taken as is from https://github.com/BabbleSim/native_simulator/
It is placed in `scripts/native_simulator`, as that `scripts` folder, despite its name, is a container of tools and other ancillary support files. (it is where the [C version of kconfiglib lived](https://github.com/zephyrproject-rtos/zephyr/tree/fcb9144dca68f98d24caf39532595555355c3c1d/scripts/kconfig) until it was replaced)

----

#### How to use it:
Just build targeting `native_sim` or `native_sim_64`, otherwise everything will look and feel like with native_posix(_64), for ex.:
```
west build -b native_sim ../tests/kernel/sched/deadline/
build/zephyr/zephyr.exe
```
You can use it with twister, for ex: `twister -p native_sim -T tests/kernel/` , 
and you can chose to build with llvm apart from gcc, and to build with a different libC:
```
ZEPHYR_TOOLCHAIN_VARIANT=llvm cmake ../samples/cpp/cpp_synchronization/ -DBOARD=native_sim -DCONFIG_PICOLIBC=y -GNinja && ninja

zephyr/zephyr.exe -stop_at=2
```

aescolar committed:
Kconfig: Add new NATIVE_LIBRARY and NATIVE_BUILD options

To differentiate builds which will produce a native executable
as direct output from the Zephyr build,
or a library which can be used later.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
Kconfig: Tidy up dependants of NATIVE_APPLICATION

So they depend or select on the right NATIVE_BUILD
instead of NATIVE_APPLICATION.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
Native simulator: Add first version in-tree

Add the first version of the native simulator.
The simultaor is taken as is from
https://github.com/BabbleSim/native_simulator/
sha: 74986abfe088a1780e604dae65f87470b4c2a0eb

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
tests POSIX: Do not filter fully by POSIX arch but by type of build

Some POSIX arch targets support now the POSIX API.
Instead of filtering the architecture fully,
let's limit it to the type of build

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
samples zbus uart_bridge: Add overlay for native_sim

Add overlay configuration for native_sim.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
tests ADC API: Add overlay for native_sim

Add DT overlay for native_sim.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
cmake: Treat native libraries differently where needed

When building a native library there is a few options
we cannot pass to the compiler and linker,
including instructing them to genrate non PIE code
(as it is still to early to say that),
or garbage collect unused sections (as we are not yet doing
the final linking).

We also need to provide different
link options when building the elf for the DTS gen_handles
parsing (as that script requires a "final" executable elf)
than when we build the native library itself.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
arch POSIX: Use posix cheats only for native apps

native libraries do not need it

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
native_sim: Add new native_sim board

Add the new native_sim board, based on the native simulator.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
arch POSIX: provide a bottom which uses natsim's thread emul

This allows building with embedded libCs in the Zephyr side,
as the POSIX arch bottom is not anymore built in Zephyr context.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
samples display lvgl: Add overlay for native_sim

Add overlays for the native_sim board

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
libC: STDOUT_CONSOLE narrow filtering by type of native build

It is possible to build with STDOUT_CONSOLE with
the embedded C libraries with the POSIX arch.
Narrow down the filtering accordingly.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
libC: PICOLIBC_SUPPORTED: Enable for NATIVE_LIBRARY

It is possible to build with the PICOLIBC_MODULE
with the POSIX arch targets which use the native
simulator as runner.
Update filtering accordingly.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
valgrind: Update suppression file to align with updated POSIX arch

Things have moved around so we need to update the
suppression file so the callstack and types of
leaks valgrind reports still match.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
drivers for POSIX arch: Add external libc dependencies

Quite a few of the drivers meant for the POSIX arch
interacted with the host directly, and will not
work when we use an embedded libC.

Until we fix them, let's add the appropriate
kconfig dependencies to avoid users trying to build them.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
MAINTAINERS: Update to cover the native simulator

Update the native posix area to cover also the native simulator,
and fix the drivers include path pattern, as some do not have the
whole native_posix in their name.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
CODEOWNERS: Cover also new native_simulator folder

Add aescolar as codeowner for the new native_simulator folder

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

ElvarT reviewed (approved):
Giving my approval, as nothing more than a humble developer using Zephyr.
I have tried this and it works really well.
My tests work just like in native_posix, now with the embedded library :)

I would like to thank the Zephyr developers for building such nice tools
to help us develop our code. These save us **lots** of time.
So, kudos, and keep up the good work.

rodrigopex reviewed (dismissed):
In a general overview, I don't feel capable of deciding if the PR can be merged. However, regarding ZBus, performing the same addition you did on the `uart_bridge` sample to the `remote_mock` one will be necessary.

aescolar committed:
samples zbus remote mock: Add overlay for native_sim

Add overlay configuration for native_sim, and enable it
for this platform in the test yaml filter.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar commented:
Thanks for taking a look @rodrigopex . 

> In a general overview, I don't feel capable of deciding if the PR can be merged.

That is ok. If you are ok with the files in your area, you can leave a review neither approving or rejecting the PR (just a "comment") stating that you are ok with the changes in your area.

> However, regarding ZBus, performing the same addition you did on the `uart_bridge` sample to the `remote_mock` one will be necessary.

I added an extra commit on top  for the remote_mock sample, with the kconfig overlay, and enabling that test for native_sim. Note though that the other changes done to samples in this PR are for those which are necessary to avoid CI breaking. Meaning for the samples which would not filter native_sim out by default. The remote_mock sample is filtering so it could have come in a follow up PR.
So far, I tried to keep this PR relatively small, and intend to follow it up with other PRs enabling more tests/samples.

rodrigopex reviewed (commented):
@aescolar, the changes regarding ZBus are ok. ✅

cfriedt commented on arch/posix/CMakeLists.txt:
```diff
@@ -16,25 +16,80 @@ endif()
+  if (CONFIG_64BIT)
+    # By default the native simulator builds and links a 32bits (x86) executable
+    target_link_options(native_simulator INTERFACE "-m64")
+    target_compile_options(native_simulator INTERFACE "-m64")
```
This will not work on Linux/Aarch64 machines (for the most part). IIRC, there is some existing logic for that which should be used.

aescolar replied:
Thanks @cfriedt, this if has been moved into the same if/else for Linux.Aarch64.

tejlmand reviewed (changes requested):
First round of review.

aescolar commented on scripts/native_simulator/Makefile:
```diff
@@ -0,0 +1,127 @@
+# Copyright 2023 Nordic Semiconductor ASA
```
The suggestion to put the native simulator here came from @carlescufi . The motivation being:
_It is placed in scripts/native_simulator, as that scripts folder, despite its name, is a container of tools and other ancillary support files. (it is where the [C version of kconfiglib lived](https://github.com/zephyrproject-rtos/zephyr/tree/fcb9144dca68f98d24caf39532595555355c3c1d/scripts/kconfig) until it was replaced)_
If you have a better suggestion you are welcome to tell.

tejlmand replied:
if it has been agreed to place it here, then I won't object to that.

Just wanted to ensure it was a conscious and agreed decision to place it here.

aescolar commented on CMakeLists.txt:
```diff
@@ -1766,15 +1782,27 @@ if(CONFIG_BUILD_OUTPUT_STRIPPED)
+      ALL
+      COMMENT "Building native simulator runner, and linking final executable"
+      COMMAND
+      make -f ${ZEPHYR_BASE}/scripts/native_simulator/Makefile all --warn-undefined-variables -r
+        NSI_CONFIG_FILE=${CMAKE_BINARY_DIR}/zephyr/NSI/nsi_config
+      DEPENDS ${logical_target_for_zephyr_elf}
```
> I'm unaware about Windows support for the simulator, 

It is not supported, and it won't be.

> This should be transformed into an external project.

Ok, let's talk.

aescolar commented on CMakeLists.txt:
```diff
@@ -41,6 +41,22 @@ endif()
 # details.
 set(CMAKE_EXECUTABLE_SUFFIX .elf)
 
+# This native_simulator library is used to pass options to the
+# native_simulator runner build. Currently the following are used:
+#  INTERFACE_COMPILE_OPTIONS:
+#      Extra compile options to be used during the build of the runner files
+#      For ex. target_compile_options(native_simulator INTERFACE "-m64")
+#  INTERFACE_LINK_OPTIONS:
+#      Extra link options to be passed during the *final* link of the runner
+#      with the embedded SW.
+#      For ex. target_link_options(native_simulator INTERFACE "-lstdc++")
+#  INTERFACE_SOURCES:
+#      Extra sources to be built in the native simulator runner context
+#      For ex. target_sources(native_simulator INTERFACE silly.c)
+#      Note that these are built with the host libC and the include directories
+#      the runner is built with.
+add_library(native_simulator INTERFACE)
```
It needs to be somewhere that is parsed very early, so other libraries can add to it.
Would you be happy if we put it in a separate file say (arch/posix/native_simulator.cmake), and we added an include here?

aescolar commented on arch/posix/core/posix_core_nsi.c:
```diff
@@ -0,0 +1,57 @@
+/*
```
Done two lines below

tejlmand commented on scripts/native_simulator/.gitignore:
```diff
@@ -0,0 +1,11 @@
+_build/*
+*.o
+*.a
+*.d
+*.elf
+*.gcda
+*.gcno
```
accepted

tejlmand commented on boards/posix/common/sdl/CMakeLists.txt:
```diff
@@ -7,6 +7,10 @@ zephyr_library_compile_definitions(NO_POSIX_CHEATS)
+if (CONFIG_NATIVE_APPLICATION)
+	zephyr_link_libraries(${SDL2_LIBRARIES})
+else()
+	target_link_options(native_simulator INTERFACE "-l${SDL2_LIBRARIES}")
```
minor nit, why using `target_link_options` when linking to a library ?

There is a `target_link_libraries(...)` available for that.
```suggestion
	target_link_libraries(native_simulator INTERFACE ${SDL2_LIBRARIES})
```

aescolar replied:
Discussed offline with @tejlmand :
`target_link_libraries()` does not seem directly/easily usable.
I will add a small comment in this cmake file in a follow up PR.
The reason for using target_link_options instead of target_link_libraries is that
target_link_libraries() updates INTERFACE_LINK_LIBRARIES but what ends up in it is something like `::@(0x56457c309b90) SDL2::@` (see https://cmake.org/cmake/help/latest/prop_tgt/INTERFACE_LINK_LIBRARIES.html#prop_tgt:INTERFACE_LINK_LIBRARIES ) which cannot be passed directly to the linker.
Right now we use target_link_options (and INTERFACE_LINK_OPTIONS) to pass and extract the links options for the runner final link.

cfriedt reviewed (approved):
LGTM as far as libc and libcpp are concerned. 

The only thing I think might be worth exploring would be a different way of encapsulating the interface(s), compile-time constant configuration, and mutable data, specific to the interface(s).

E.g. everything in the global namespace, vs using a struct, function pointers, struct for compile-time-const config data and another struct for mutable data (minimizing scope / residuals).

Mentioned [here](https://github.com/zephyrproject-rtos/zephyr/pull/59662#issuecomment-1620031992) as well. 

Related it to our device model because that's fairly mature now and it's a pretty decent means of abstracting most hardware and platform things.

Definitely excited for this work though - thanks @aescolar !
